### PR TITLE
Don't require changes to read_table_A when creating new device handlers

### PR DESCRIPTION
**Pull request by bpmartin20**
TL;DR: Allow read_table_A to dynamically load modules for new kinds of devices or objects, without needing any code changes in read_table_A.

Currently, read_table_A needs to be updated every time a new device handler (e.g. "type", the first field of a record) is developed. This creates a maintenance burden in any of the following cases:

Someone develops a new handler:
- for a one-off device they built
- that is based on proprietary  information that can't be published
- that takes a long time to develop (because we have busy lives and other things claiming our attention)
- that otherwise doesn't make sense to publish to the rest of the world. 

In any these scenarios, they are forced to keep maintaining their own read_table_A module, somewhat independently of the upstream repository. Git with rebase helps to an extent, but this requires more knowledge of git and generally is an unnecessary burden.

This PR eliminates the need to update read_table_A for each new device handler. It amends the way that read_table_A  handles unrecognized types. Instead of simply throwing an error, it first attempts to locate a module matching the lowercase of the .mht type field (first field). If found, read_table_A loads that module and writes code into the .mhp file to invoke its "new" method. The code passes all the parameters found on .mht record for the new method to use in setting up the object. An optional "init" method is subsequently called if present, to handle any post-instantiation activities, like adding an object to a group. In this way our developer can simply create a module with the proper name and leave read_table_A unaltered.

### CODE REVIEW

**Review by marcmerlin — COMMENTED** [review]
@hollie  @hplato  this is absolutely the kind of change that I like, but I don't honestly know the mh code enough and haven't looked at it in 10 years, to be qualified to review this.
Are you able to review this better than me?
If not, are you willing to let this in with a message to the mailing list asking people to try this master branch and make sure they don't get broken by the change?

**Comment by bpmartin20** [comment]
I'm not interested in trying to lobby anybody for this change. I trust you folks to make good decisions in that regard. Perhaps I can propose a way to more easily assess the risk, though.

Basically, there are two changes here...

1) The heart of this change is to look for a module if the first field of the .mht record isn't recognized. Previously that condition triggered an error message. So this part will only come into play in what is currently an error condition. If there's a bug in the new code, it'll only come into play on installations that are already broken. I think it'll be pretty easy to see when the code triggers, and if you agree, any unexpected impact should be on a very small number of installations.

2) I took a block of unnamed code that assessed group membership, and made it into a subroutine. That's a little more invasive, though turning a code block into a subroutine is usually pretty easy to do. Between the two, I suggest this is the area that demands the most scrutiny. If you can assess whether I converted the block into a subroutine successfully, that should help you form an opinion regarding that code. Bonus points if you detect where I fixed a subtle bug in which the group membership count could get incremented even if the item doesn't get added to the group. The bug probably never triggered in the wild, but now it can't.

And by all means, post to the mailing list about the new master. I think a call for testing is always important with new masters.

Thank you, everyone.

**Comment by bpmartin20** [comment]
I'll be off the grid through the end of April starting Tuesday. Is there anything you need from me before I disappear?

**Comment by hplato** [comment]
I really like the idea behind this change, the only concern is that it adds a lot of functionality which will need to be tested. With updates to MH slowing, I think adding in user contributions is great, and this might be a release that lasts a long time.

So unless anyone has any objections, I’d say accept it, and with the HA_Items updates, we have a strong and updated base master to test and validate in May with a summer release (I’ve been working to test HA_Item and migrate my devices. I’ll look at a release after that).

thoughts?

**Comment by marcmerlin** [comment]
@hplato  as per my last message, I have pretty much 0 time to work on this, but I think in this case it's ok to accept this and post to the list asking for people to test master and report back if it broke them.
Even better, tell them to check master now, test it, then a few days later, merge this and tell them to test again.
